src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,567 @@
+                auto uncore_freq = static_cast<uint64_t>(*it);
+                if (std::isnan(mapped_mem_bw)) {
+                    throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                    "(): mapped CPU_UNCORE_FREQUENCY with no max memory bandwidth.",
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+                // Just make sure the frequency does not have multiple definitions.
+                if (!policy_uncore_freqs.insert(uncore_freq).second) {
+                    throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                    " policy has multiple entries for CPU_UNCORE_FREQUENCY " +
+                                    std::to_string(uncore_freq),
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+            }
+            else if (!std::isnan(mapped_mem_bw)) {
+                throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                " policy maps a NaN CPU_UNCORE_FREQUENCY with max memory bandwidth: " +
+                                std::to_string(mapped_mem_bw),
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+
+            }
+        }
+
+    }
+
+    // Distribute incoming policy to children
+    void CPUActivityAgent::split_policy(const std::vector<double>& in_policy,
+                                    std::vector<std::vector<double> >& out_policy)
+    {
+        for (auto &child_pol : out_policy) {
+            child_pol = in_policy;
+        }
+    }
+
+    // Indicate whether to send the policy down to children
+    bool CPUActivityAgent::do_send_policy(void) const
+    {
+        return true;
+    }
+
+    void CPUActivityAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                                        std::vector<double>& out_sample)
+    {
+
+    }
+
+    // Indicate whether to send samples up to the parent
+    bool CPUActivityAgent::do_send_sample(void) const
+    {
+        return false;
+    }
+
+    void CPUActivityAgent::adjust_platform(const std::vector<double>& in_policy)
+    {
+        m_do_write_batch = false;
+
+        if (m_update_qm_max_rate) {
+            for (auto it = in_policy.begin() + M_POLICY_FIRST_UNCORE_FREQ;
+                 it != in_policy.end() && std::next(it) != in_policy.end();
+                 std::advance(it, 2)) {
+
+                if (!std::isnan(*it)) {
+                    auto uncore_freq = static_cast<uint64_t>(*it);
+                    auto max_mem_bw = *(it + 1);
+                    // Not valid to have NAN max mem bw for uncore freq.
+                    GEOPM_DEBUG_ASSERT(!std::isnan(max_mem_bw),
+                                       "mapped CPU_UNCORE_FREQUENCY with no max memory bandwidth assigned.");
+                    m_qm_max_rate[uncore_freq] = max_mem_bw;
```

Comment:
This seems like it can be moved into validate_policy(), then the if() block on 298 down to 320 can be removed from here.